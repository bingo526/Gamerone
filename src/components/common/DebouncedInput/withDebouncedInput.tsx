import React from 'react';
import debounce from 'lib/debounce';

export interface Props extends React.HTMLAttributes<HTMLInputElement> {
  value?: string | null;
  inputRef: React.Ref<HTMLInputElement>;
}

type DebouncedInputState = {
  value?: string | null;
};

export default function withDebouncedInput(
  WrappedComponent: any,
  config = { timeout: 500 },
) {
  return class DebouncedInput extends React.Component<Props> {
    state: DebouncedInputState;

    constructor(props: Props) {
      super(props);
      this.state = {
        value: this.props.value === null ? '' : this.props.value,
      };

      this.sendTextChange = debounce(this.sendTextChange, config.timeout);
    }

    static getDerivedStateFromProps(nextProps: Props) {
      return nextProps.value === null ? { value: '' } : null;
    }

    handleTextChange = (e: any) => {
      this.setState({ value: e.target.value });
      this.sendTextChange({ target: { value: e.target.value } });
    };

    sendTextChange = (e: any) => {
      if (this.props.onChange) this.props.onChange(e);
    };

    render() {
      return (
        <WrappedComponent
          {...this.props}
          value={this.state.value} // todo vadim
          ref={this.props.inputRef}
          onChange={this.handleTextChange.bind(this)}
        />
      );
    }
  };
}
