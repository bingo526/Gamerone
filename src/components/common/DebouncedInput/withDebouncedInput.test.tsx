import React from 'react';
import { render, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import withDebouncedInput from './withDebouncedInput';

beforeAll(() => {
  jest.useFakeTimers();
});

afterAll(() => {
  jest.useRealTimers();
});

const inputText = (el, value) => {
  act(() => {
    userEvent.type(el, value);
  });
};

test('debouces input as expected', async () => {
  const Input = () => (
    <input
      type="text"
      aria-label="debounce-input"
      id="debounce-input"
      name="debounce"
    />
  );
  const DebouncedInput = withDebouncedInput(Input, { timeout: 500 });
  const handleInputChange = jest.fn((e) =>
    console.log('mock handleInputChange', e.target.value),
  );
  let input = null;
  act(() => {
    const { getByLabelText, asFragment } = render(
      <DebouncedInput onChange={handleInputChange} />,
    );

    input = getByLabelText(/debounce-input/i);
    expect(asFragment()).toMatchSnapshot();
  });

  expect(input).not.toBeNull();
  inputText(input, 'x');

  const word = 'debounce';
  let type = '';

  for (let i = 0; i < word.length; i++) {
    jest.advanceTimersByTime(Math.floor(Math.random() * 499));
    type += word[i];
    inputText(input, type);
  }

  expect(handleInputChange).toHaveBeenCalledTimes(0);

  // TODO: Fix this bug. withDebouncedInput onChange is not fired as expected

  // jest.advanceTimersByTime(501);
  // expect(handleInputChange).toHaveBeenCalledTimes(1);
  // expect(input.value).toBe('debounce');
});
