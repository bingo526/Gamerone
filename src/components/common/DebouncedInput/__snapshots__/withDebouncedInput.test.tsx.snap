// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`debouces input as expected 1`] = `
<DocumentFragment>
  <input
    aria-label="debounce-input"
    id="debounce-input"
    name="debounce"
    type="text"
  />
</DocumentFragment>
`;
